# Plotting Exercises, Part 1

><span style="color: #9370DB">*Bárbara Flores*</span>

### Exercise 1

Create a pandas dataframe from the "Datasaurus.txt" file using the code: 

Note that the file being downloaded is *not* actually a CSV file. It is tab-delimited, meaning that within each row, columns are separated by tabs rather than commas. We communicate this to pandas with the `delimiter="\t"` option (`"\t"` is how we write a tab, as we will discuss in future lessons). 

In [1]:
import pandas as pd

pd.set_option("mode.copy_on_write", True)

# Download the data
datasaurus = pd.read_csv(
    "https://raw.githubusercontent.com/nickeubank/practicaldatascience/master"
    "/Example_Data/Datasaurus.txt",
    delimiter="\t",
)

datasaurus.sample(5)

example1_x  example1_y  example2_x  example2_y  example3_x  example3_y  \
134   42.476792   56.731290   68.995029   13.994311   60.787970   17.046205   
37    21.778040    6.616625   41.389523   37.296056   39.485575   20.033050   
69    62.934459   31.916391   53.101382   84.547659   51.473459   79.159867   
110   48.823197   76.635331   52.992029    9.906668   64.010241   17.924478   
16    40.570197   79.372963   36.894328   47.706556   26.163387   62.026815   

     example4_x  example4_y  example5_x  example5_y  ...  example9_x  \
134     68.4615      5.2564   75.448063   17.045149  ...   53.512971   
37      35.1282     30.2564   25.855128   17.081909  ...   36.619722   
69      56.6667     86.0256   50.464225   82.935328  ...   51.308825   
110     58.7179     18.3333   50.699851   17.085184  ...   68.412767   
16      32.8205     51.0256   27.338869   51.653047  ...   31.135735   

     example9_y  example10_x  example10_y  example11_x  example11_y  \
134    5.875110    77.573153    14.780646    69.501045     5.955834   
37    36.811414    41.640454    27.093095    30.503474    38.348556   
69    80.136391    58.221613    90.411019    50.373519    82.219206   
110   19.913127    76.731849    16.216846    50.377402    16.506015   
16    61.431913    29.285147    63.383606    30.494421    45.773476   

     example12_x  example12_y  example13_x  example13_y  
134    66.816914    12.273294    80.573273     8.365639  
37     31.090065    32.383755    40.915972    23.541592  
69     64.935929    87.072138    41.787323    83.988078  
110    63.946952    18.692845    42.826756    25.556959  
16     32.438650    49.903019    48.664925    64.961005  

[5 rows x 26 columns]

### Exercise 2

This dataset actually contains 13 separate example datasets, each with two variables named `example[number]_x` and `example[number]_y`. 

In order to get a better sense of what these datasets look like, write a loop that iterates over each example dataset (numbered 1 to 13) and print out the mean and standard deviation for `example[number]_x` and `example[number]_y` for each dataset. 

For example, the first iteration of this loop might return something like:

```
Example Dataset 1: 
Mean x: 23.12321978429576, 
Mean y: 98.23980921730972, 
Std Dev x: 21.2389710287, 
Std Dev y: 32.2389081209832,
Correlation: 0.73892819281
```

(Though you shouldn't get those specific values)

In [2]:
for i in range(13):
    x = datasaurus[datasaurus.columns[i * 2]]
    y = datasaurus[datasaurus.columns[i * 2 + 1]]
    x_mean = x.mean()
    y_mean = y.mean()
    x_std = x.std()
    y_std = y.std()
    corr = x.corr(y)
    print(f"Example Dataset {i+1}:")
    print(f"Mean x: {x_mean:4f},")
    print(f"Mean y: {y_mean:4f},")
    print(f"Std Dev x: {x_std:4f},")
    print(f"Std Dev y: {y_std:4f},")
    print(f"Correlation: {corr:4f}\n")

Example Dataset 1:
Mean x: 54.266100,
Mean y: 47.834721,
Std Dev x: 16.769825,
Std Dev y: 26.939743,
Correlation: -0.064128

Example Dataset 2:
Mean x: 54.268730,
Mean y: 47.830823,
Std Dev x: 16.769239,
Std Dev y: 26.935727,
Correlation: -0.068586

Example Dataset 3:
Mean x: 54.267320,
Mean y: 47.837717,
Std Dev x: 16.760013,
Std Dev y: 26.930036,
Correlation: -0.068343

Example Dataset 4:
Mean x: 54.263273,
Mean y: 47.832253,
Std Dev x: 16.765142,
Std Dev y: 26.935403,
Correlation: -0.064472

Example Dataset 5:
Mean x: 54.260303,
Mean y: 47.839829,
Std Dev x: 16.767735,
Std Dev y: 26.930192,
Correlation: -0.060341

Example Dataset 6:
Mean x: 54.261442,
Mean y: 47.830252,
Std Dev x: 16.765898,
Std Dev y: 26.939876,
Correlation: -0.061715

Example Dataset 7:
Mean x: 54.268805,
Mean y: 47.835450,
Std Dev x: 16.766704,
Std Dev y: 26.939998,
Correlation: -0.068504

Example Dataset 8:
Mean x: 54.267849,
Mean y: 47.835896,
Std Dev x: 16.766759,
Std Dev y: 26.936105,
Correlation: -0.068980



### Exercise 3

Based only on these results, discuss what might you conclude about these example datasets with your partner. Write down your thoughts.

><span style="color: #9370DB">*When analyzing the statistics at first glance, they seem to be quite similar. The mean of x is around 54.26, and the mean of Y is 47.8. The same holds for the standard deviation; among the different examples, the data appears to be quite similar.*</span>
>
><span style="color: #9370DB">*It's probably necessary to visualize the data to observe if there are different patterns.*</span>

### Execise 4

Write a loop that iterates over these example datasets, and using Altair library, plot a simple scatter plot of each dataset with the `x` variable on the x-axis and the `y` variable on the y-axis.

Hint: When writing this type of code, it is often best to start by writing code to do what you want for the first iteration of the loop. Once you have code that works for the first example dataset, then write the full loop around it. 

Hint 2: To force Jupyter to display your charts when they're generated within a loop, use the method `.show()` (e.g. `my_chart.show()`).

Hint 3: You will need to change the range of the axes to make the plots look good!

In [3]:
import altair as alt

for i in range(13):
    df = datasaurus[datasaurus.columns[i * 2 : i * 2 + 2]]
    x_name = df.columns[0]
    y_name = df.columns[1]
    chart = alt.Chart(df).mark_circle().encode(x=x_name, y=y_name)

    chart.display()

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

### Exercise 5

Review you plots. How does your impression of how these datasets differ from what you wrote down in Exercise 3?

><span style="color: #9370DB">*At first glance, when we only analyze the statistics such as mean, standard deviation, and correlation, all datasets appear to be very similar. However, when plotted, we realize that they are distributed in very different ways, information that we couldn't identify before graphing.*</span>
>
><span style="color: #9370DB">*This is a small example of the powerful tool that data visualization represents*</span>

## Economic Development and... Your Choice!

### Exercise 6

Load the World Development Indicator data used in the [plotting reading](https://nickeubank.github.io/practicaldatascience_book/notebooks/class_5/week_2/20_plotting_with_seaborn.html). Rather than picking a single year, pick a single country and look at how GDP per capita and one of the other variables in that dataset have evolved together over time. 

Make any adjustments to the functional forms of your variables and/or axes needed to make the figure legible. 

### Exercise 7

Now add a second series. Facet your plot so that the two subplots are positioned so that they are effectively sharing the same time axes (e.g., if you draw a line up from 2010 on one plot, you get to 2010 on the other).

Rather than telling you exactly how to do it, however, I'll point you to the [seaborn tutorial](https://seaborn.pydata.org/tutorial/objects_interface#faceting-and-pairing-subplots). It has examples that don't do exactly what you want, but should be close enough you can guess-and-check to the solution you want!

Use your detective skills (and some guess and check work) to figure out how to get it to work!
